lsp/src/protocol.ml

```diff
@@ -3391,6 +3406,482 @@ module Hover = struct
   [@@@end]
 end
 
+module SignatureHelpOptions = struct
+  type t = { triggerCharacters : string list [@default []] }
+  [@@deriving_inline yojson]
+
+  let _ = fun (_ : t) -> ()
+
+  let t_of_yojson =
+    ( let _tp_loc = "lsp/src/protocol.ml.SignatureHelpOptions.t" in
+      function
+      | `Assoc field_yojsons as yojson -> (
+        let triggerCharacters_field = ref None
+        and duplicates = ref []
+        and extra = ref [] in
+        let rec iter = function
+          | (field_name, _field_yojson) :: tail ->
+            ( match field_name with
+            | "triggerCharacters" -> (
+              match Ppx_yojson_conv_lib.( ! ) triggerCharacters_field with
+              | None ->
+                let fvalue = list_of_yojson string_of_yojson _field_yojson in
+                triggerCharacters_field := Some fvalue
+              | Some _ ->
+                duplicates := field_name :: Ppx_yojson_conv_lib.( ! ) duplicates
+              )
+            | _ ->
+              if
+                Ppx_yojson_conv_lib.( ! )
+                  Ppx_yojson_conv_lib.Yojson_conv.record_check_extra_fields
+              then
+                extra := field_name :: Ppx_yojson_conv_lib.( ! ) extra
+              else
+                () );
+            iter tail
+          | [] -> ()
+        in
+        iter field_yojsons;
+        match Ppx_yojson_conv_lib.( ! ) duplicates with
+        | _ :: _ ->
+          Ppx_yojson_conv_lib.Yojson_conv_error.record_duplicate_fields _tp_loc
+            (Ppx_yojson_conv_lib.( ! ) duplicates)
+            yojson
+        | [] -> (
+          match Ppx_yojson_conv_lib.( ! ) extra with
+          | _ :: _ ->
+            Ppx_yojson_conv_lib.Yojson_conv_error.record_extra_fields _tp_loc
+              (Ppx_yojson_conv_lib.( ! ) extra)
+              yojson
+          | [] ->
+            let triggerCharacters_value =
+              Ppx_yojson_conv_lib.( ! ) triggerCharacters_field
+            in
+            { triggerCharacters =
+                ( match triggerCharacters_value with
+                | None -> []
+                | Some v -> v )
+            } ) )
+      | _ as yojson ->
+        Ppx_yojson_conv_lib.Yojson_conv_error.record_list_instead_atom _tp_loc
+          yojson
+      : Ppx_yojson_conv_lib.Yojson.Safe.t -> t )
+
+  let _ = t_of_yojson
+
+  let yojson_of_t =
+    ( function
+      | { triggerCharacters = v_triggerCharacters } ->
+        let bnds : (string * Ppx_yojson_conv_lib.Yojson.Safe.t) list = [] in
+        let bnds =
+          let arg = yojson_of_list yojson_of_string v_triggerCharacters in
+          ("triggerCharacters", arg) :: bnds
+        in
+        `Assoc bnds
+      : t -> Ppx_yojson_conv_lib.Yojson.Safe.t )
+
+  let _ = yojson_of_t
+
+  [@@@end]
+end
+
+module ParameterInformation = struct
+  type t =
+    { label : (int * int) Or_string.t
```

Comment:
nit: I'd have another datatype here for label, protocol says:

> Either a string or an inclusive start and exclusive end offsets within its containing signature label.

I think having something like:
```
type paramInfoLabel = 
  | BySubstring of string
  | ByRange of int * int` would
```
would be self-documenting, otherwise I'd expect code which uses those types to have comments describing what means what.